zebra/zebra_rib.c

```diff
@@ -2976,49 +3033,137 @@ void rib_update_table(struct route_table *table, rib_update_event_t event)
 		 * has already been queued  we don't
 		 * need to queue it up again
 		 */
-		if (rn->info && CHECK_FLAG(rib_dest_from_rnode(rn)->flags,
-					   RIB_ROUTE_ANY_QUEUED))
+		if (rn->info
+		    && CHECK_FLAG(rib_dest_from_rnode(rn)->flags,
+				  RIB_ROUTE_ANY_QUEUED))
 			continue;
+
 		switch (event) {
+		case RIB_UPDATE_KERNEL:
+			rib_update_route_node(rn, ZEBRA_ROUTE_KERNEL);
+			break;
 		case RIB_UPDATE_RMAP_CHANGE:
 		case RIB_UPDATE_OTHER:
-			/* Right now, examine all routes. Can restrict to a
-			 * protocol in
-			 * some cases (TODO).
-			 */
-			if (rnode_to_ribs(rn)) {
-				RNODE_FOREACH_RE_SAFE (rn, re, next)
-					SET_FLAG(re->status,
-						 ROUTE_ENTRY_CHANGED);
-				rib_queue_add(rn);
-			}
+			rib_update_route_node(rn, ZEBRA_ROUTE_ALL);
 			break;
-
 		default:
 			break;
 		}
 	}
 }
 
-/* RIB update function. */
-void rib_update(vrf_id_t vrf_id, rib_update_event_t event)
+static void rib_update_handle_vrf(vrf_id_t vrf_id, rib_update_event_t event)
 {
 	struct route_table *table;
 
+	if (IS_ZEBRA_DEBUG_EVENT)
+		zlog_debug("%s: Handling VRF %s event %s", __func__,
+			   vrf_id_to_name(vrf_id), rib_update_event2str(event));
+
 	/* Process routes of interested address-families. */
 	table = zebra_vrf_table(AFI_IP, SAFI_UNICAST, vrf_id);
-	if (table) {
-		if (IS_ZEBRA_DEBUG_EVENT)
-			zlog_debug("%s : AFI_IP event %d", __func__, event);
+	if (table)
 		rib_update_table(table, event);
-	}
 
 	table = zebra_vrf_table(AFI_IP6, SAFI_UNICAST, vrf_id);
-	if (table) {
-		if (IS_ZEBRA_DEBUG_EVENT)
-			zlog_debug("%s : AFI_IP6 event %d", __func__, event);
+	if (table)
 		rib_update_table(table, event);
-	}
+}
+
+static void rib_update_handle_vrf_all(rib_update_event_t event)
+{
+	struct zebra_router_table *zrt;
+
+	if (IS_ZEBRA_DEBUG_EVENT)
+		zlog_debug("%s: Handling VRF (ALL) event %s", __func__,
+			   rib_update_event2str(event));
+
+	/* Just iterate over all the route tables, rather than vrf lookups */
+	RB_FOREACH (zrt, zebra_router_table_head, &zrouter.tables)
+		rib_update_table(zrt->table, event);
+}
+
+struct rib_update_ctx {
+	rib_update_event_t event;
+	bool vrf_all;
+	vrf_id_t vrf_id;
+};
+
+static struct rib_update_ctx *rib_update_ctx_init(vrf_id_t vrf_id,
+						  rib_update_event_t event)
+{
+	struct rib_update_ctx *ctx;
+
+	ctx = XCALLOC(MTYPE_RIB_UPDATE_CTX, sizeof(struct rib_update_ctx));
+
+	ctx->event = event;
+	ctx->vrf_id = vrf_id;
+
+	return ctx;
+}
+
+static void rib_update_ctx_fini(struct rib_update_ctx *ctx)
```

Comment:
I like the dtor pattern - but could it take a `**`, and NULL the `ctx` pointer so it can't be stale after the fini call?

Reply:
will do